tests/unit/services/test_config.py

```diff
@@ -0,0 +1,223 @@
+# This file is part of craft-application.
+#
+# Copyright 2024 Canonical Ltd.
+#
+# This program is free software: you can redistribute it and/or modify it
+# under the terms of the GNU Lesser General Public License version 3, as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful, but WITHOUT
+# ANY WARRANTY; without even the implied warranties of MERCHANTABILITY,
+# SATISFACTORY QUALITY, or FITNESS FOR A PARTICULAR PURPOSE.
+# See the GNU Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+"""Unit tests for the configuration service."""
+
+import itertools
+import json
+import string
+import subprocess
+from collections.abc import Iterator
+
+import craft_application
+import craft_cli
+import pytest
+import pytest_subprocess
+from craft_application.services import config
+from hypothesis import given, strategies
+
+CRAFT_APPLICATION_TEST_ENTRY_VALUES = [
+    *(
+        ("verbosity_level", mode.name.lower(), mode)
+        for mode in craft_cli.messages.EmitterMode
+    ),
+    *(("verbosity_level", mode.name, mode) for mode in craft_cli.messages.EmitterMode),
+    ("debug", "true", True),
+    ("debug", "false", False),
+    ("build_environment", "host", "host"),
+    ("secrets", "Cara likes Butterscotch.", "Cara likes Butterscotch."),
+    ("platform", "laptop", "laptop"),
+    ("platform", "mainframe", "mainframe"),
+    ("build_for", "riscv64", "riscv64"),
+    ("build_for", "s390x", "s390x"),
+    *(("parallel_build_count", str(i), i) for i in range(10)),
+    *(("max_parallel_build_count", str(i), i) for i in range(10)),
+]
+APP_SPECIFIC_TEST_ENTRY_VALUES = [
+    ("my_str", "some string", "some string"),
+    ("my_int", "1", 1),
+    ("my_int", "2", 2),
+    ("my_bool", "true", True),
+    ("my_bool", "false", False),
+    ("my_default_str", "something", "something"),
+    ("my_default_int", "4294967296", 2**32),
+    ("my_bool", "1", True),
+]
+TEST_ENTRY_VALUES = CRAFT_APPLICATION_TEST_ENTRY_VALUES + APP_SPECIFIC_TEST_ENTRY_VALUES
+
+
+@pytest.fixture(scope="module")
+def app_environment_handler(default_app_metadata) -> config.AppEnvironmentHandler:
+    return config.AppEnvironmentHandler(default_app_metadata)
+
+
+@pytest.fixture(scope="module")
+def craft_environment_handler(default_app_metadata) -> config.CraftEnvironmentHandler:
+    return config.CraftEnvironmentHandler(default_app_metadata)
+
+
+@pytest.fixture(scope="module")
+def snap_config_handler(default_app_metadata) -> Iterator[config.SnapConfigHandler]:
+    with pytest.MonkeyPatch.context() as monkeypatch:
+        monkeypatch.setenv("SNAP", "/snap/testcraft/x1")
+        monkeypatch.setenv("SNAP_COMMON", "/")
+        monkeypatch.setenv("SNAP_DATA", "/")
```

Comment:
Can we ask for key through the specific handler?
If I can see correctly, now if we create `data` config key in `CRAFT` / `APP` it will always supersede the `SNAP_` environment variables.
Do we need to handle such case?

Reply:
See above - the ConfigurationService doesn't get `SNAP_*` environment variables.